Centralize dependency versions for client libraries

bsiegel opened the pull request:
The management libraries will be tackled in a separate PR.

Commit:
Centralize dependency versions

weshaggard commented on sdk/eventhub/Microsoft.Azure.EventHubs.Processor/src/Microsoft.Azure.EventHubs.Processor.csproj:
```diff
@@ -17,17 +17,17 @@
   <ItemGroup>
-    <PackageReference Include="Newtonsoft.Json" Version="12.0.1" />
-    <PackageReference Include="Microsoft.Azure.KeyVault.Core" Version="3.0.3" />
+    <PackageReference Include="Newtonsoft.Json" VersionOverride="12.0.1" />
```
For this one lets try and pull it down the 10.0.3

weshaggard commented on Packages.props:
```diff
@@ -0,0 +1,108 @@
+  <ItemGroup Condition=" '$(TargetFramework)' == 'netstandard2.0' ">
+    <PackageReference Update="Microsoft.AspNetCore.WebUtilities" Version="2.2.0" />
+  </ItemGroup>
+  <!-- <ItemGroup Condition="'$(IsTargetingNetFx)' == 'true'">
```
Let's set Newtonsoft.Json to version 10.0.3 as the global default and add overrides if they are needed. I think we should try to move the 12.0.1 back to 10.0.3 but there may be some cases were we need to stick to 6.0.8 for the older target frameworks but that would be an override if needed.

weshaggard commented on Packages.props:
```diff
@@ -0,0 +1,108 @@
+    <PackageReference Update="Microsoft.Extensions.Logging.Abstractions" Version="2.2.0" />
+    <PackageReference Update="Microsoft.Extensions.Options" Version="2.2.0" />
+    <PackageReference Update="Microsoft.Extensions.PlatformAbstractions" Version="1.1.0" />
+    <!-- <PackageReference Update="Microsoft.IdentityModel.Clients.ActiveDirectory" Version="3.14.0" /> -->
```
Let's try aligning everyone to the most common version which appears to be 4.5.1.

weshaggard commented on Packages.props:
```diff
@@ -0,0 +1,108 @@
+    <PackageReference Update="System.Threading.Tasks.Extensions" Version="4.5.1" />
+    <PackageReference Update="System.ValueTuple" Version="4.5.0" />
+    <PackageReference Update="WindowsAzure.ServiceBus" Version="5.1.0" />
+    <!-- <PackageReference Update="WindowsAzure.Storage" Version="8.1.4" /> -->
```
Lets try to align everyone to 8.1.4.

weshaggard commented:
Looks like EventHub seems to be the most common outlier in the version differences lets try and use the common version there as well and see if we can get it working and if not we consider an override or upgrade.

Commit:
Align packages that differ on major versions

Microsoft.IdentityModel.Clients.ActiveDirectory -> 4.5.1
Newtonsoft.Json -> 10.0.3
WindowsAzure.Storage -> 8.1.4

weshaggard commented on Packages.props:
```diff
@@ -0,0 +1,103 @@
+<?xml version="1.0" encoding="utf-8"?>
```
I would like to try and keep the root as clean as possible so prefer this to go into the eng directory.  According to the docs at https://github.com/Microsoft/MSBuildSdks/tree/master/src/CentralPackageVersions you can set the `CentralPackagesFile` in the props file to put the packages file somewhere else.

Commit:
Fix build issues after dependency version alignment

Batch & keyvault tests use our Microsoft.Rest.ClientRuntime.Azure
libraries which themselves require an ancient version of the ADAL.

Eventhubs requires WindowsAzure.Storage >= 9.0, which seems like it
should be backwards compatible with 8.x which was used in Batch &
HDinsights.

Commit:
Fix xunit tests

mitchdenny commented:
Large cross repo commit == abuse errors going off ... looks like we still need the token caching solution to go live :)

Commit:
Conditionally compile out CertificateCredentialsNetCore

bsiegel commented:
/azp run net - template - ci

bsiegel commented:
/azp run net - storage - ci

bsiegel commented:
/azp run net - servicebus - ci

bsiegel commented:
@weshaggard Everything seems to be working here with the exception of a single failing eventhubs test. I have done a bit of searching and I'm not sure how to fix this - any suggestions? The failure is:

```
Failed   DisposeAsyncDelegatesToCloseAsync

Error Message:
 System.IO.FileLoadException : Could not load file or assembly 'System.Runtime.CompilerServices.Unsafe, Version=4.0.4.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' or one of its dependencies. The located assembly's manifest definition does not match the assembly reference. (Exception from HRESULT: 0x80131040)
Stack Trace:
   at System.Threading.Tasks.ValueTask.get_IsCompleted()
   at System.Runtime.CompilerServices.ValueTaskAwaiter.get_IsCompleted()
   at Azure.Messaging.EventHubs.Tests.EventHubClientTests.<DisposeAsyncDelegatesToCloseAsync>d__25.MoveNext() in D:\a\1\s\sdk\eventhub\Azure.Messaging.EventHubs\tests\EventHubClient\EventHubClientTests.cs:line 537
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at NUnit.Framework.Internal.TaskAwaitAdapter.GenericAdapter`1.GetResult()
   at NUnit.Framework.Internal.AsyncToSyncAdapter.Await(Func`1 invoke)
   at NUnit.Framework.Internal.Commands.TestMethodCommand.RunTestMethod(TestExecutionContext context)
   at NUnit.Framework.Internal.Commands.TestMethodCommand.Execute(TestExecutionContext context)
   at NUnit.Framework.Internal.Execution.SimpleWorkItem.PerformWork()
```

weshaggard commented on eng/Packages.Data.props:
```diff
@@ -0,0 +1,103 @@
+    <PackageReference Update="System.Security.Cryptography.Cng" Version="4.3.0" />
+    <PackageReference Update="System.Security.Cryptography.Primitives" Version="4.3.0" />
+    <PackageReference Update="System.Text.Json" Version="4.6.0-preview6.19259.10" />
+    <PackageReference Update="System.Threading.Tasks.Extensions" Version="4.5.1" />
```
I believe this is what is causing your test failure. You need to bump System.Threading.Tasks.Extensions to 4.5.2.

Commit:
Attempt to fix test failure on Windows

bsiegel commented:
/azp run net - keyvault - ci

bsiegel commented:
/azp run net - operationalinsights - ci

bsiegel commented:
/azp run net - servicebus - ci

bsiegel commented:
In case anyone comes in from the internet searching for that exception stack trace above, the solution is that the following libraries must all use the **SAME** version (4.5.1 or 4.5.2, it most likely doesn't matter as long as they all use the same version):

- System.Memory
- System.Runtime.CompilerServices.Unsafe
- System.Threading.Tasks.Extensions
